tools/testbench/common_test_ipc4.c

```diff
@@ -0,0 +1,593 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2018-2024 Intel Corporation. All rights reserved.
+
+#if CONFIG_IPC_MAJOR_4
+
+#include <sof/audio/component_ext.h>
+#include <sof/lib/notifier.h>
+#include <sof/audio/component_ext.h>
+#include <sof/schedule/edf_schedule.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <ipc/stream.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "testbench/common_test.h"
+#include "testbench/file.h"
+#include "testbench/topology_ipc4.h"
+
+#if defined __XCC__
+#include <xtensa/tie/xt_timer.h>
+#endif
+
+SOF_DEFINE_REG_UUID(testbench);
+DECLARE_TR_CTX(testbench_tr, SOF_UUID(testbench_uuid), LOG_LEVEL_INFO);
+LOG_MODULE_REGISTER(testbench, CONFIG_SOF_LOG_LEVEL);
+
+/* testbench helper functions for pipeline setup and trigger */
+
+int tb_setup(struct sof *sof, struct testbench_prm *tp)
+{
+	struct ll_schedule_domain domain = {0};
+	int bits;
+	int krate;
+
+	domain.next_tick = tp->tick_period_us;
+
+	/* init components */
+	sys_comp_init(sof);
+
+	/* Module adapter components */
+	sys_comp_module_crossover_interface_init();
+	sys_comp_module_dcblock_interface_init();
+	sys_comp_module_demux_interface_init();
+	sys_comp_module_drc_interface_init();
+	sys_comp_module_eq_fir_interface_init();
+	sys_comp_module_eq_iir_interface_init();
+	sys_comp_module_file_interface_init();
+	sys_comp_module_gain_interface_init();
+	sys_comp_module_google_rtc_audio_processing_interface_init();
+	sys_comp_module_igo_nr_interface_init();
+	sys_comp_module_mfcc_interface_init();
+	sys_comp_module_multiband_drc_interface_init();
+	sys_comp_module_mux_interface_init();
+	sys_comp_module_rtnr_interface_init();
+	sys_comp_module_selector_interface_init();
+	sys_comp_module_src_interface_init();
+	sys_comp_module_asrc_interface_init();
+	sys_comp_module_tdfb_interface_init();
+	sys_comp_module_volume_interface_init();
```

Comment:
can we not setup the linker script to do dynamic init?

Reply:
@cujomalainey just had a chat with @singalsu on this, it looks like the C runtime for the simulator is quite basic wrt POSIX APIs i.e. no dl_open() and simulation target needs to be an executable. It looks like we have to do this and state this in a comment.

Reply:
bleh, thanks